sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala

```diff
@@ -0,0 +1,149 @@
+    if (!containsAccountInfo(ethTransaction.getFrom)) {
+      log.error(s"Adding transaction failed because mempoolMap is not initialized for account ${ethTransaction.getFrom}")
+      throw new IllegalStateException(s"MempoolMap is not initialized for account ${ethTransaction.getFrom}")
+    }
+
+    if (!contains(ethTransaction)) {
+
+      val account = ethTransaction.getFrom
+      val expectedNonce = nonces.get(account).get
+      if (expectedNonce.equals(ethTransaction.getNonce)) {
+        all.put(ethTransaction.id, ethTransaction)
+        val executableTxsPerAccount = executableTxs.get(account).getOrElse({
+          val txsPerAccountMap = new mutable.TreeMap[BigInteger, ModifierId]()
+          executableTxs.put(account, txsPerAccountMap)
+          txsPerAccountMap
+        })
+        executableTxsPerAccount.put(ethTransaction.getNonce, ethTransaction.id)
+        var nextNonce = expectedNonce.add(BigInteger.ONE)
+        nonExecutableTxs.get(account).foreach(txs => {
+          while (txs.contains(nextNonce)) {
+            val promotedTxId = txs.remove(nextNonce).get
+            executableTxsPerAccount.put(nextNonce, promotedTxId)
+            nextNonce = nextNonce.add(BigInteger.ONE)
+          }
+          if (txs.isEmpty) {
+            nonExecutableTxs.remove(account)
+          }
+        }
+
+        )
+        nonces.put(account, nextNonce)
+
+      }
+      else if (expectedNonce.compareTo(ethTransaction.getNonce) < 0) {
+        val listOfTxs = nonExecutableTxs.get(account).getOrElse({
+          val txsPerAccountMap = new mutable.TreeMap[BigInteger, ModifierId]()
+          nonExecutableTxs.put(account, txsPerAccountMap)
+          txsPerAccountMap
+        })
+        if (listOfTxs.contains(ethTransaction.getNonce)) {
+          val oldTxId = listOfTxs.get(ethTransaction.getNonce).get
+          val oldTx = all.get(oldTxId).get
+          if (canPayHigherFee(ethTransaction, oldTx)) {
+            log.debug(s"Replacing transaction ${oldTx} with ${ethTransaction}")
+            all.remove(oldTxId)
+            all.put(ethTransaction.id, ethTransaction)
+            listOfTxs.put(ethTransaction.getNonce, ethTransaction.id)
+          }
+        }
+        else {
+          all.put(ethTransaction.id, ethTransaction)
+          listOfTxs.put(ethTransaction.getNonce, ethTransaction.id)
+        }
+      }
+      else {
+        val listOfTxs = executableTxs.get(account).get
+        val oldTxId = listOfTxs.get(ethTransaction.getNonce).get
+        val oldTx = all.get(oldTxId).get
+        if (canPayHigherFee(ethTransaction, oldTx)) {
+          log.debug(s"Replacing transaction ${oldTx} with ${ethTransaction}")
+          all.remove(oldTxId)
+          all.put(ethTransaction.id, ethTransaction)
+          listOfTxs.put(ethTransaction.getNonce, ethTransaction.id)
+        }
+
+      }
+    }
+    this
+  }
+
+  def canPayHigherFee(newTx: SidechainTypes#SCAT, oldTx: SidechainTypes#SCAT): Boolean = {
```

Comment:
I have moved canPayHigherFeeThan declaration in AccountTransaction  but I haven't moved getMaxFeePerGas() / getMaxPriorityFeePerGas(), because I don't know if maxGasFee and maxPriorityFee are properties applicable to a generic AccountTransaction.

Reply:
After talking with Sasha, I have added the gas fee and gas tip methods to AccountTransaction and moved canPayHigherFee in MempoolMap.